docs/datagen/server/recipes.md

```diff
@@ -0,0 +1,185 @@
+Recipe Generation
+=================
+
+Recipes can be generated for a mod by subclassing `RecipeProvider` and overriding `#buildCraftingRecipes`. A recipe is supplied for data generation once a `FinishedRecipe` view is accepted by the consumer. `FinishedRecipe`s can either be created and supplied manually or, for convenience, created using a `RecipeBuilder`.
+
+After implementation, the provider must be [added][datagen] to the `DataGenerator`.
+
+`RecipeBuilder`
+---------------
+
+`RecipeBuilder` is a convenience implementation for creating `FinishedRecipe`s to generate. It provides basic definitions for unlocking, grouping, saving, and getting the result of a recipe. This is done through `#unlockedBy`, `#group`, `#save`, and `#getResult` respectively.
+
+!!! important
+    [`ItemStack` outputs][stack] in recipes is not supported within vanilla recipe builders for `RecipeProvider`. A `FinishedRecipe` must be built in a different manner for existing vanilla recipe serializers to generate this data.
+
+!!! warning
+    The item results being generated must have a valid `CreativeModeTab` specified; otherwise, a `NullPointerException` will be thrown. If no category accurately represents the item and no tab should be created, set the `#tab` property during item initialization to `CreativeModeTab#TAB_SEARCH`.
```

Comment:
It is the only category that can view all items within the creative inventory. As such, if no item is applicable for any tab and the user doesn't want to create a custom tab, it would make sense that it would be found in at least that location, hence the setting.

Reply:
Are there any possibly unwanted side-effects with using the search tab?

Reply:
No, creative tabs are populated using `Item#fillItemCategory` which performs a check called `#allowedIn` which determines whether or not the item can appear in the creative inventory. If the creative tab is not null, then it will populate with the item if the tab is equal to the set category or the search tab. As such, the only thing this would affect is that the OR check for creative tab would essentially be the same.

You can verify this is a workspace as well.

Reply:
Hmmm, looking at the code which generates the recipes shows that the output location of the generated recipe JSON uses the creative mode tab of the resulting item for convenient grouping: `new ResourceLocation(recipeId.getNamespace(), "recipes/" + this.result.getItemCategory().getRecipeFolderName() + "/" + recipeId.getPath())`.

This means that a recipe whose result item's tab is set to the `SEARCH` creative mode tab will be placed under `recipes/search/`, which may be an unwanted side-effect.

Reply:
Unwanted or not, you cannot generate a recipe using the builder without a creative tab, which I agree should be the case. If it gets generated into that folder as a result, it would make sense that it would belong there.

Reply:
Is it good practice to add items to the search tab, such that the item can only appear in the creative menu through the search tab and no other tab? Are there cases in which a mod has an item which does not belong on normal creative tabs (whether vanilla or modded), but they wish to add a crafting recipe for?

Reply:
This is specifically dealing with a case where a modder willingly decides to hide their item from the creative inventory but makes it obtainable in survival, so this is my opinion on how to handle this unrecommended edge case. As for the second point, I address that in the [PR that attempts to change that behavior](https://github.com/MinecraftForge/MinecraftForge/pull/8511#issuecomment-1066870648).